File: Subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        '''
        1. return all possible subsets
        2. nums' integers are unique
        3. must not contain duplicate subsets
        '''
        result = []
        
        def dfs(index, path):
            result.append(path) #every path is subset. 
                                #so append path to the result everytime to be called  
            
            for i in range(index, len(nums)): #make a selection for subset
                dfs(i+1, path + [nums[i]]) #next selection will be done from the next element, not the current one.
                
        dfs(0, []) #call dfs with the first choice
        return result

File: Subsets.java
class Solution {
    
    List<List<Integer>> res;
    public List<List<Integer>> subsets(int[] nums) {
        res = new ArrayList<List<Integer>>();
        subsetHelper(nums, 0, new ArrayList<Integer>());
        return res;
    }
    
    public void subsetHelper(int[] nums, int d, ArrayList<Integer> subSet){
        if(d == nums.length){
            res.add(new ArrayList(subSet));
        }
        else{
            subSet.add(nums[d++]);
            subsetHelper(nums, d, subSet);
            subSet.remove(subSet.size()-1);
            subsetHelper(nums, d, subSet);
        }
    }
}
//O(N * 2^N)
// class Solution {
    
//     List<List<Integer>> res;
//     public List<List<Integer>> subsets(int[] nums) {
//         res = new ArrayList<List<Integer>>();
//         subsetHelper(nums, new int[nums.length], 0);
//         return res;
//     }
    
//     public void subsetHelper(int[] nums, int[] check, int depth){
//         if(depth == nums.length){
//             ArrayList<Integer> subset = new ArrayList<>();
//             for(int i=0;i<check.length;i++){
//                 if(check[i] == 1){
//                     subset.add(nums[i]);
//                 }
//             }
//             res.add(subset);
//         }
//         else{
//             check[depth] = 0;
//             subsetHelper(nums, check, depth + 1);
//             check[depth] = 1;
//             subsetHelper(nums, check, depth + 1);
//         }
//     }
// }
